src/EnergyPlus/HVACManager.cc

```diff
@@ -1117,6 +1117,76 @@ void SimHVAC(EnergyPlusData &state)
+                        }
+                    } // CO2 not converged
+
+                    auto &arrayRef6 = state.dataConvergeParams->AirLoopConvergence(AirSysNum).HVACGenContamNotConverged;
+                    if (std::any_of(std::begin(arrayRef6), std::end(arrayRef6), [](bool i) { return i; })) {
+
+                        ShowContinueError(state,
+                                          format("Air System Named = {} did not converge for generic contaminant",
+                                                 state.dataAirLoop->AirToZoneNodeInfo(AirSysNum).AirLoopName));
+                        ShowContinueError(state, "Check values should be zero. Most Recent values listed first.");
+                        std::string HistoryTrace = "";
+                        for (int StackDepth = 0; StackDepth < DataConvergParams::ConvergLogStackDepth; ++StackDepth) {
+                            HistoryTrace += format(
+                                "{:.6R},", state.dataConvergeParams->AirLoopConvergence(AirSysNum).HVACGenContamDemandToSupplyTolValue[StackDepth]);
+                        }
+                        ShowContinueError(
+                            state, format("Demand-to-Supply interface generic contaminant check value iteration history trace: {}", HistoryTrace));
+                        HistoryTrace = "";
+                        for (int StackDepth = 0; StackDepth < DataConvergParams::ConvergLogStackDepth; ++StackDepth) {
+                            HistoryTrace +=
+                                format("{:.6R},",
+                                       state.dataConvergeParams->AirLoopConvergence(AirSysNum).HVACGenContamSupplyDeck1ToDemandTolValue[StackDepth]);
+                        }
+                        ShowContinueError(
+                            state,
+                            format("Supply-to-demand interface deck 1 generic contaminant check value iteration history trace: {}", HistoryTrace));
+
+                        if (state.dataAirLoop->AirToZoneNodeInfo(AirSysNum).NumSupplyNodes >= 2) {
+                            HistoryTrace = "";
+                            for (int StackDepth = 0; StackDepth < DataConvergParams::ConvergLogStackDepth; ++StackDepth) {
+                                HistoryTrace += format(
+                                    "{:.6R},",
+                                    state.dataConvergeParams->AirLoopConvergence(AirSysNum).HVACGenContamSupplyDeck1ToDemandTolValue[StackDepth]);
+                            }
+                            ShowContinueError(state,
+                                              format("Supply-to-demand interface deck 2 generic contaminant check value iteration history trace: {}",
+                                                     HistoryTrace));
+                        }
+                    } // Generic contaminant not converged
+
```

Comment:
Provide error messages if convergence is not met.

Reply:
Could all this repeated stuff be put into a single function and called with an array reference or two passed in?